sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CollapseEmptyPlan.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.DeclarativeAggregate
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules._
+
+/**
+ * Collapse plans consisting empty local relations generated by [[PruneFilters]].
+ * 1. Binary(or Higher)-node Logical Plans
+ *    - Union with all empty children.
+ *    - Join with one or two empty children (including Intersect/Except).
+ * 2. Unary-node Logical Plans
+ *    - Project/Filter/Sample/Join/Limit/Repartition with all empty children.
+ *    - Aggregate with all empty children and without DeclarativeAggregate expressions like COUNT.
+ *    - Generate(Explode) with all empty children. Others like Hive UDTF may return results.
+ */
+object CollapseEmptyPlan extends Rule[LogicalPlan] with PredicateHelper {
+  private def isEmptyLocalRelation(plan: LogicalPlan): Boolean = plan match {
+    case p: LocalRelation => p.data.isEmpty
+    case _ => false
+  }
+
+  private def isDeclarativeAggregate(e: Expression): Boolean = e match {
+    case _: DeclarativeAggregate => true
+    case _: LeafExpression => false
+    case other => other.children.forall(isDeclarativeAggregate)
+  }
+
+  private def empty(plan: LogicalPlan) = LocalRelation(plan.output, data = Seq.empty)
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case p: Union if p.children.nonEmpty && p.children.forall(isEmptyLocalRelation) =>
+      empty(p)
+
+    case p @ Join(_, _, joinType, _) if p.children.exists(isEmptyLocalRelation) => joinType match {
+      case Inner => empty(p)
+      case LeftOuter | LeftSemi | LeftAnti if isEmptyLocalRelation(p.left) => empty(p)
+      case RightOuter if isEmptyLocalRelation(p.right) => empty(p)
+      case FullOuter if p.children.forall(isEmptyLocalRelation) => empty(p)
+      case _ => p
+    }
```

Comment:
Could you please comment that `Intersect` is also covered here? I didn't realized that we've already translated `Intersect` using joins at first.